src/ucp/rndv/rndv.c

```diff
@@ -1426,6 +1495,162 @@ static void ucp_rndv_do_rkey_ptr(ucp_request_t *rndv_req, ucp_request_t *rreq,
+                           ucs_memory_type_t local_mem_type,
+                           ucp_rkey_h *rkey_p)
+{
+    ucp_context_h context                    = worker->context;
+    ucp_md_index_t rkey_ptr_md_index         = UCP_NULL_RESOURCE;
+    ucp_rkey_h rkey                          = NULL;
+    ucp_rndv_ppln_rkey_ptr_data_t *ppln_data = NULL;
+    int found                                = 0;
+    ucp_md_map_t rkey_ptr_md_map;
+    ucp_md_index_t md_index;
+    const uct_md_attr_t *md_attr;
+    ucp_ep_h mem_type_ep;
+    ucp_lane_index_t mem_type_rma_lane;
+    unsigned rkey_index;
+    void *local_ptr;
+    khiter_t iter;
+    int ret;
+    ucs_status_t status;
+
+    if ((local_mem_type == UCS_MEMORY_TYPE_HOST) ||
+        (local_mem_type == UCS_MEMORY_TYPE_UNKNOWN)) {
+        return NULL;
+    }
+
+    mem_type_ep = worker->mem_type_ep[local_mem_type];
+    if (mem_type_ep == NULL) {
+        return NULL;
+    }
+
+    ucs_for_each_bit(md_index, ucp_rkey_packed_md_map(rtr_hdr + 1)) {
+        md_attr = &context->tl_mds[md_index].attr;
+        if ((md_attr->cap.flags & UCT_MD_FLAG_RKEY_PTR) &&
+            /* Do not use xpmem, because cuda_copy registration will fail and
+             * performance will not be optimal. */
+            !(md_attr->cap.flags & UCT_MD_FLAG_REG) &&
+            (md_attr->cap.access_mem_types & UCS_BIT(remote_mem_type))) {
+            rkey_ptr_md_index = md_index;
+            break;
+        }
+    }
+
+    if (rkey_ptr_md_index == UCP_NULL_RESOURCE) {
+        return NULL;
+    }
+
+    rkey_ptr_md_map = UCS_BIT(rkey_ptr_md_index);
+
+    if (ucs_unlikely(ucp_ep_ext_control(ep)->peer_mem == NULL)) {
+        ucp_ep_ext_control(ep)->peer_mem = kh_init(ucp_ep_peer_mem_hash);
+        iter = kh_end(ucp_ep_ext_control(ep)->peer_mem);
+    } else {
+        iter = kh_get(ucp_ep_peer_mem_hash, ucp_ep_ext_control(ep)->peer_mem,
+                      rtr_hdr->address);
+        if (iter != kh_end(ucp_ep_ext_control(ep)->peer_mem)) {
+            ppln_data = &kh_val(ucp_ep_ext_control(ep)->peer_mem, iter);
+            if (rtr_hdr->size <= ppln_data->size) {
+                rkey  = ppln_data->rkey;
+                found = 1;
+            } else {
+                ucp_rndv_ppln_data_destroy(context, ppln_data);
+            }
+        }
+    }
+
+    if (iter == kh_end(ucp_ep_ext_control(ep)->peer_mem)) {
+        iter = kh_put(ucp_ep_peer_mem_hash, ucp_ep_ext_control(ep)->peer_mem,
+                      rtr_hdr->address, &ret);
+        ucs_assertv_always((ret == UCS_KH_PUT_BUCKET_EMPTY) ||
+                           (ret == UCS_KH_PUT_BUCKET_CLEAR), "ret %d", ret);
+        ppln_data = &kh_val(ucp_ep_ext_control(ep)->peer_mem, iter);
+    }
+
+    ucs_assert(ppln_data != NULL);
+
+    if (!found) {
+        ucp_ep_rkey_unpack_md_map(ep, rtr_hdr + 1, 0, rkey_ptr_md_map, &rkey);
+        ppln_data->rkey = rkey;
+        ppln_data->size = rtr_hdr->size;
+    }
```

Comment:
this part seems overly complicated.
1. need to extract to separate function to lookup peer_mem and create it on demand, recreate if size is smaller, etc
2. use only kh_put, not kg_get+kh_put

Reply:
Will kh_put realloc or have more overhead even if found in the hash?
IMO it's worth it to simplify the code.
At least (like wrote in other comment) should try get rid of "found" variable